struct.js

```diff
@@ -0,0 +1,260 @@
+    self.annotations = def.annotations;
+    self.valueType = def.valueType;
+    self.constructDefaultValue = null; // def.defaultValue
+    self.to = null;
+}
+
+FieldSpec.prototype._link = function link(spec) {
+    var self = this;
+    self.to = spec._getSpecByDefinition(self.valueType);
+};
+
+function StructSpec() {
+    var self = this;
+    self.name = null;
+    self.fields = [];
+    self.fieldNames = [];
+    self.fieldsById = {};
+    self.fieldsByName = {};
+    self.Constructor = null;
+    self.RW = new StructRW(self);
+}
+
+StructSpec.prototype.typeid = 12;
+
+StructSpec.prototype._process = function process(def) {
+    var self = this;
+    self.name = def.annotations && def.annotations['js.name'] || def.name;
+    var fields = def.fields;
+    for (var index = 0; index < fields.length; index++) {
+        var fieldDef = fields[index];
+        var field = new FieldSpec(fieldDef);
+        field.name = field.annotations && field.annotations['js.name'] || field.name;
+        self.fieldsById[field.id] = field;
+        self.fieldsByName[field.name] = field;
+        self.fieldNames[index] = field.name;
+        self.fields.push(field);
+    }
+};
+
+StructSpec.prototype._link = function link(spec) {
+    var self = this;
+    self.Constructor = createConstructor(self.name, self.fieldNames);
+    self.Constructor.RW = self.RW;
+    spec[self.name] = self.Constructor;
+    for (var index = 0; index < self.fields.length; index++) {
+        self.fields[index]._link(spec);
+    }
+};
+
+function createConstructor(name, fieldNames) {
+    var source;
+    source = '(function ' + name + '(options) {\n';
+    for (var index = 0; index < fieldNames.length; index++) {
+        var fieldName = fieldNames[index];
+        source += '    this.' + fieldName + ' = null;\n';
+        source += 'if (options && \'' + fieldName + '\' in options) ' +
+            '{ this.' + fieldName + ' = options.' + fieldName + '; }';
+    }
+    source += '})\n';
+    // eval is an operator that captures the lexical scope of the calling
+    // function and deoptimizes the lexical scope.
+    // By using eval in an expression context, it loses this second-class
+    // capability and becomes a first-class function.
+    // (0, eval) is one way to use eval in an expression context.
+    return (0, eval)(source);
+}
+
+function StructRW(spec) {
+    assert(spec, 'spec required');
+    var self = this;
+    self.spec = spec;
+}
+
+StructRW.prototype.byteLength = function byteLength(struct) {
+    var self = this;
+    var spec = self.spec;
+    var length = 1; // stop:1
+    var result;
+    for (var index = 0; index < spec.fields.length; index++) {
+        var field = spec.fields[index];
+        var value = struct[field.name];
+        var available = value !== null && value !== undefined;
+
+        // TODO deal with varying default optional or required semantics based
+        // on whether this is a result or argument struct.
```

Comment:
We have options. @raynos favors throwing an error if a field does not declare that it is either optional or required. A reasonable default would be a long bike-shed. Also, the presence of a defaultValue implies optional.